### ISSUE TITLE
[META] Symfony 6 return type declarations

### ISSUE THREAD

**Issue by wouterj** [symfony/symfony#43021]
Symfony 6 comes (almost) fully typed. If you're an open source package maintainer, adding a return type might cause BC breaks (as it forces your users to add the type as well if they implement or override the method).
We might consider, _on a case by case basis_, whether it's worth to postpone adding a type declaration for a specific method to Symfony 7. This might help you, as a maintainer, to release a new minor instead of a new major version to add support for Symfony 6.

1. In order to properly judge this, we ask you to **install 5.4-dev in your package and run the [type patch utility](https://symfony.com/doc/5.4/setup/upgrade_major.html#upgrading-to-symfony-6-add-native-return-types)** twice (also temporarily install the `symfony/error-handler` component if it's not available already):

   ```
   $ SYMFONY_PATCH_TYPE_DECLARATIONS=force=1 ./vendor/bin/patch-type-declarations
   $ SYMFONY_PATCH_TYPE_DECLARATIONS=force=1 ./vendor/bin/patch-type-declarations
   ```

2. This script patches any methods that are private, final, `@internal` or tests. **Please double check if the unpatched methods should not be final or internal** (i.e. should a user really extend or use this class?).

3. If the _second_ run **produces any deprecations in the output for methods that must be considered public/protected, please copy paste them as a comment to this issue** (including a link to the package). E.g:

   > https://github.com/wouterj/WouterJEloquentBundle
   > ```
   > WouterJ\EloquentBundle\Command\SeedCommand
   > Method "Symfony\Component\Console\Command\Command::execute()" might add "int" as a native return type declaration in the future. Do the same in child class "WouterJ\EloquentBundle\Command\SeedCommand" now to avoid errors or add an explicit @return annotation to suppress this message.
   > ```

   Note that the current 6.0 branch reflects which types are not going to be included in that release. These still produce deprecations in 5.4+, which can be resolved by adding the `@return` PHPdoc.

   (please only post deprecations mentioning classes from Symfony, other deprecations should be reported in the relevant package issue tracker)

We don't promise anything, but we might postpone some types based on the feedback in *this* issue. For instance, we already did for the `Command::execute(): int` method shown in the example.

**Comment by jordisala1991** [comment]
https://github.com/sonata-project/SonataMediaBundle

https://github.com/sonata-project/SonataMediaBundle/blob/master/src/Form/DataTransformer/ProviderDataTransformer.php
```
Sonata\MediaBundle\Form\DataTransformer\ProviderDataTransformer
Method "Symfony\Component\Form\DataTransformerInterface::transform()" might add "mixed" as a native return type declaration in the future. Do the same in implementation "Sonata\MediaBundle\Form\DataTransformer\ProviderDataTransformer" now to avoid errors or add an explicit @return annotation to suppress this message.
```

```
Sonata\MediaBundle\Form\DataTransformer\ProviderDataTransformer
Method "Symfony\Component\Form\DataTransformerInterface::reverseTransform()" might add "mixed" as a native return type declaration in the future. Do the same in implementation "Sonata\MediaBundle\Form\DataTransformer\ProviderDataTransformer" now to avoid errors or add an explicit @return annotation to suppress this message.
```

The problem with that is the mixed return type was added on php8, If sonata bundles are still supporting php 7.4 o 7.3 we can't add the mixed return. I saw that you didn't add that return type on 6.0 yet, do you plan to do it or it will be delayed to 7.0?

**Comment by wouterj** [comment]
@jordisala1991 thanks for the feedback! Yes, that return type won't be introduced in 6.0. Any missing type from the current 6.0 branch is not planned for the stable release (and see also https://github.com/symfony/symfony/blob/6.0/.github/expected-missing-return-types.diff ).
So you can suppress it using `@return mixed` for now.

**Comment by ddaffie** [comment]
@wouterj Thank you for asking for feedback. I am working in the Drupal core project and we are preparing for using Symfony 6 in our next major release. We are getting the following deprecation warning: "Method "Symfony\Component\Serializer\Normalizer\NormalizerInterface::normalize()" will return "array|string|int|float|bool|\ArrayObject|null" as of its next major version." The problem for us is that we have the method return an object. We have now 3 options: rewrite the normalize system to return an allowed value, or detach it from the Symfony method or ask if the return type hint can be changed to "mixed". Changing the return type hint to "mixed" would save us a lot of time. Our issue for the problem is: https://www.drupal.org/project/drupal/issues/3232074.

**Comment by jordisala1991** [comment]
https://github.com/KnpLabs/KnpMenuBundle
https://github.com/KnpLabs/KnpMenuBundle/blob/master/src/Templating/Helper/MenuHelper.php
```
Knp\Bundle\MenuBundle\Templating\Helper\MenuHelper
Method "Symfony\Component\Templating\Helper\HelperInterface::getName()" might add "string" as a native return type declaration in the future. Do the same in implementation "Knp\Bundle\MenuBundle\Templating\Helper\MenuHelper" now to avoid errors or add an explicit @return annotation to suppress this message.
```

I think this is the only one thats prevents KnpMenuBundle (Sonata also depends on this) from working on Symfony 6.

**Comment by derrabus** [comment]
> The problem for us is that we have the method return an object.

What kind of object is this? The problem is that the allowed return types define what kind of input the encoders can expect. It's a contract between normalizers and encoders. That is why it has been narrowed down to what _all_ encoders understand.

**Comment by wouterj** [comment]
@jordisala1991 I think it makes sense to revert the native return type and instead use a PHPdoc return type for Symfony 6 on that method. Are you open to create a PR to 6.0 doing that? Thanks!

**Comment by jordisala1991** [comment]
I will! Tonight or maybe tomorrow

**Comment by bbrala** [comment]
@derrabus while thinking about your comment one thing came to mind, since php8+ is required, shouldn't that interface support [Stringable](https://www.php.net/manual/en/class.stringable.php)? 

The awnser about the returned object required a little more typing, which i can't do right now :)

**Comment by derrabus** [comment]
> shouldn't that interface support [Stringable](https://www.php.net/manual/en/class.stringable.php)?

No, because casting a stringable object to a string should be the concern of the normalizer, not the encoder.

**Comment by wouterj** [comment]
Thanks @jordisala1991!

_Fyi, I've hidden all comments as "resolved" when we've resolved that return type. This way, I hope this issue can be kept a bit organized. This has nothing to do with the validity of the comments._

**Comment by dunglas** [comment]
There is a problem with `@return` annotations and union types

For instance, if the interface has `@return void|object`, the script adds `: void` to the method instead of `: mixed` or nothing.

Reproducer: https://github.com/api-platform/core/blob/2.6/src/Bridge/Doctrine/Common/DataPersister.php#L50

**Comment by dunglas** [comment]
We identified another problem while running the tool on the API Platform codebase. When using the tool with the `force=phpdoc` mode, sometimes annotations aren't inserted at a good place and create syntax errors:

Exemple:

```
// ...
         */
        public function parseLiteral(/*Node */ $valueNode, ?array $variables = null)
     * @return mixed
        {
```

Reproducer: run `SYMFONY_PATCH_TYPE_EXCLUDE='/Uuid|NelmioApiDoc|ArraySubsetLegacy|DataPersister/' SYMFONY_PATCH_TYPE_DECLARATIONS='force=phpdoc&php=7.1' ./vendor/bin/patch-type-declarations` on https://github.com/api-platform/core/issues/4495.

**Comment by kbond** [comment]
[`Symfony\Component\Console\Input\InputInterface`](https://github.com/symfony/symfony/blob/1d7bc05b8e5867d14299b93e528941878f13388a/src/Symfony/Component/Console/Input/InputInterface.php#L82) is proving troublesome to support in 7.4 and 8.0 (the `mixed` returns).

**Comment by derrabus** [comment]
@kbond Do you maintain a package that provides its own implementations of that interface?

**Comment by kbond** [comment]
It is an implementation in a _yet to be released_ package I'm working on - I was planning on supporting 7.4 & 8.0.

**Comment by derrabus** [comment]
Can you provide a PR against 6.0 that removes the return types that are in your way? Maybe it's better to discuss this on actual code.

**Comment by mbabker** [comment]
Can we re-evaluate the return typehint on `Symfony\Component\Serializer\Normalizer\NormalizerInterface::normalize()`?  I finally was able to unblock some of my own dependencies and get 6.0 components installed, and the one change that is proving to be the most painful in everything I've worked on so far is that typehint. 
 https://github.com/BabDev/PagerfantaBundle/commit/6983c668ee5294183a428dd41121a28633772417 is actually the final working version that I landed on without dropping PHP 7.4 support (there was a simpler version in [this commit](https://github.com/BabDev/PagerfantaBundle/commit/92d90bca7ba5621c926a7aa18f884d60898d6b08) but [CI failed](https://github.com/BabDev/PagerfantaBundle/actions/runs/1339706645) with parse errors on the PHP 7.4 builds).

**Comment by bbrala** [comment]
I'd like to add how we use the `normalize()` function in Drupal and why it is as it is. I also have a possible solution that might be usefull.

I'll open a PR that targets the 6.0 branch so we can explore the possible implications.

**Comment by derrabus** [comment]
@mbabker If you revert that commit and simply add an `array` return type to the `normalize()` method – wouldn't that class work with Symfony 5 and 6?

**Comment by mbabker** [comment]
> @mbabker If you revert that commit and simply add an `array` return type to the `normalize()` method – wouldn't that class work with Symfony 5 and 6?

Huh, TIL PHP's type variance will allow that - https://3v4l.org/E1Rng

**Comment by franmomu** [comment]
Hi, while trying to test [SncRedisBundle with Symfony 6](https://github.com/snc/SncRedisBundle/pull/630), an error appeared [in CI](https://github.com/snc/SncRedisBundle/runs/4019689740?check_suite_focus=true) because of adding return type declarations to 

https://github.com/symfony/symfony/blob/563dbf66b6b28b52d2ff9f5c5721619c87130225/src/Symfony/Component/HttpFoundation/Session/Storage/Handler/AbstractSessionHandler.php#L41-L45

is it ok to relax return type declaration on those methods?

**Comment by stof** [comment]
@franmomu you can add the string return type in the child class even when running on Symfpny 5 (on PHP 7.2+). That variance rule of PHP is precisely what allows adding return types in Symfony 6 while packages can support Symfony 5 and 6.

**Comment by franmomu** [comment]
> @franmomu you can add the string return type in the child class even when running on Symfpny 5 (on PHP 7.2+). That variance rule of PHP is precisely what allows adding return types in Symfony 6 while packages can support Symfony 5 and 6.

Sorry, my explanation was vague, I was referring to BC break for someone extending the `RedisSessionHandler` from that bundle if we add return types declarations in a minor version.

**Comment by stof** [comment]
Well, you indeed need to do a major version if you want to add the return type in a class that is not final (which is why Symfony does those additions in Symfony 6). However, that's totally something that a bundle could do (and if we say that bundles must be able to upgrade to Symfony 6 in minor versions, then we must cancel the whole typing initiative, as there is no way to guarantee that a bundle extending one of our classes does that in a final class only)

**Comment by derrabus** [comment]
`AbstractSessionHandler` is one of those classes where we had to add return types to some methods becuase PHP 8.1 urges us to do so. This of course does not affect the methods you've described, but I don't think we do ourselves a favor if we rollback the return types just for those methods and keep them on other methods of the very same class.

**Comment by franmomu** [comment]
> However, that's totally something that a bundle could do (and if we say that bundles must be able to upgrade to Symfony 6 in minor versions, then we must cancel the whole typing initiative, as there is no way to guarantee that a bundle extending one of our classes does that in a final class only)

yeah, that's why I asked, because I think this issue is about that:

> Symfony 6 comes (almost) fully typed. If you're an open source package maintainer, adding a return type might cause BC breaks (as it forces your users to add the type as well if they implement or override the method).
We might consider, on a case by case basis, whether it's worth to postpone adding a type declaration for a specific method to Symfony 7. This might help you, as a maintainer, to release a new minor instead of a new major version to add support for Symfony 6.

anyway, we'll go with the minor BC break, thanks.

**Comment by stof** [comment]
@franmomu for the case of the SessionHandler, I think it is not worth delaying the type addition. There are very few bundles that implement custom handlers, and SncRedisBundle could release a major version IMO (if we delay it, the bundle will still need to release a major version at some point, as it would still happen later)

**Comment by wouterj** [comment]
Thanks for sharing the finding it in this issue! This kind of discussion is exactly what this issue is about, even if we ultimately decide not to loosen the type.

**Comment by jordisala1991** [comment]
For Symfony Panther I am trying the following PR: https://github.com/symfony/panther/pull/509

As you can see, there is a commit (https://github.com/symfony/panther/pull/509/commits/31d379ed84cc7cb041ce2e4ebedeeda448d0a2b3) with the problematic changes, I will list them here so you can decide wether is worth it removing them (I can do the PR). They are more related to the fact that Symfony Panther currently supports from PHP 7.2 and Symfony 6 is forcing me to add return typehints only compatible with PHP 8.

Here is the list:

```
Symfony\Component\Panther\Client
Method "Symfony\Component\BrowserKit\AbstractBrowser::getServerParameter()" might add "mixed" as a native return type declaration in the future. Do the same in implementation "Symfony\Component\Panther\Client" now to avoid errors or add an explicit @return annotation to suppress this message.
```

```
Symfony\Component\Panther\DomCrawler\Field\ChoiceFormField
Method "Symfony\Component\DomCrawler\Field\FormField::getValue()" might add "array|string|null" as a native return type declaration in the future. Do the same in implementation "Symfony\Component\Panther\DomCrawler\Field\ChoiceFormField" now to avoid errors or add an explicit @return annotation to suppress this message.
```

```
Symfony\Component\Panther\DomCrawler\Field\FileFormField
Method "Symfony\Component\DomCrawler\Field\FormField::getValue()" might add "array|string|null" as a native return type declaration in the future. Do the same in implementation "Symfony\Component\Panther\DomCrawler\Field\FileFormField" now to avoid errors or add an explicit @return annotation to suppress this message.
```

```
Symfony\Component\Panther\DomCrawler\Field\FormFieldTrait
Method "Symfony\Component\DomCrawler\Field\FormField::getValue()" might add "array|string|null" as a native return type declaration in the future. Do the same in implementation "Symfony\Component\Panther\DomCrawler\Field\FormFieldTrait" now to avoid errors or add an explicit @return annotation to suppress this message.
```

```
Symfony\Component\Panther\DomCrawler\Form
Method "Symfony\Component\DomCrawler\Form::get()" might add "FormField|array" as a native return type declaration in the future. Do the same in implementation "Symfony\Component\Panther\DomCrawler\Form" now to avoid errors or add an explicit @return annotation to suppress this message.
```

```
Symfony\Component\Panther\DomCrawler\Form
Method "Symfony\Component\DomCrawler\Form::offsetGet()" might add "FormField|array" as a native return type declaration in the future. Do the same in implementation "Symfony\Component\Panther\DomCrawler\Form" now to avoid errors or add an explicit @return annotation to suppress this message.
```

There are 4 methods that need a typehint removal (one of them affect 3 classes for Panther)
Another thing that I don't know if has to be fixed here or in Symfony Panther is this change:

`createClient` in SymfonyPanther is returning `AbstractBrowser`: https://github.com/symfony/panther/blob/main/src/WebTestAssertionsTrait.php#L280
`createClient` in Symfony is returning `KernelBrowser`: https://github.com/symfony/symfony/blob/6.0/src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php#L38

**Comment by nicolas-grekas** [comment]
Now closing as resolved, thanks to everyone involved!
